### DISCUSSION TITLE
Procedural mesh with large number of triangles

### DISCUSSION THREAD

**Discussion by ghost** [bevyengine/bevy#5032] [Q&A] [answered]
Hey there ! I'm thinking about starting a toy project in Bevy, it would involve creating several procedurally generated meshes that can change their topology at runtime. I think I will need around 30 of these meshes with at least 130k triangles each. Is it something that is possible to make with Bevy ? Would regenerating one of such meshes at runtime stall the game ?

**Comment by HackerFoo** [comment, marked as answer]
It's possible. I've made something like that for mobile devices with Bevy: https://noumenal.app
If you can generate the mesh (vertex positions, normals, etc.) fast enough, Bevy should be able to render them.

If it takes more than a frame to generate a mesh, you can use an [asynchronous task](https://github.com/bevyengine/bevy/blob/latest/examples/async_tasks/async_compute.rs) or a separate thread and an `Arc<Mutex<T>>` resource.

**Reply by ghost** [reply]
I have a similar project set up on Unreal Engine and while I can generate vertex positions on an off-thread, the function that creates a mesh from the vertices must be called from the game thread and if the number of vertices is large - it will stall the game. Is it the case for Bevy as well ?

**Reply by HackerFoo** [reply]
Both app logic and the renderer are multithreaded in Bevy. Loading a `Mesh` object of that size is pretty quick. It should be easy to write a benchmark that generates a sphere with as many faces as you want and loads it each frame, and you can use Tracy to look at the timing on your target machine.
